### PR TITLE
🌱 Implement Cluster TopologyReconciled v1beta2 condition

### PR DESCRIPTION
**Pull request by sbueringer**
Signed-off-by: Stefan Büringer buringerst@vmware.com

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Part of #11105

### CODE REVIEW

File: api/v1beta1/cluster_types.go [hunk 1]
```diff
@@ -56,6 +56,59 @@ const (
...
+	// ClusterTopologyReconcileSucceededV1Beta2Reason documents the reconciliation of a Cluster topology succeeded.
+	ClusterTopologyReconcileSucceededV1Beta2Reason = "TopologyReconcileSucceeded"
+
+	// ClusterTopologyReconciledFailedV1Beta2Reason documents the reconciliation of a Cluster topology
+	// failing due to an error.
+	ClusterTopologyReconciledFailedV1Beta2Reason = "TopologyReconcileFailed"
+
+	// ClusterTopologyReconciledControlPlaneUpgradePendingV1Beta2Reason documents reconciliation of a Cluster topology
+	// not yet completed because Control Plane is not yet updated to match the desired topology spec.
+	ClusterTopologyReconciledControlPlaneUpgradePendingV1Beta2Reason = "ControlPlaneUpgradePending"
```
Comment (fabriziopandini) [inline api/v1beta1/cluster_types.go:69]: I think we should discuss which of those reason might impact cluster availability, and if not, improve the custom merge strategy in the Cluster controller. My first take is that TopologyReconcileFailed, ClusterClassNotReconciled are the only two reasons that should impact availability, everything else should be considered info

File: internal/controllers/topology/cluster/conditions.go [hunk 2]
```diff
@@ -103,6 +148,13 @@ func (r *Reconciler) reconcileTopologyReconciledCondition(s *scope.Scope, cluste
 				s.HookResponseTracker.AggregateMessage(),
 			),
 		)
+		v1beta2conditions.Set(cluster, metav1.Condition{
+			Type:   clusterv1.ClusterTopologyReconciledV1Beta2Condition,
+			Status: metav1.ConditionFalse,
+			Reason: clusterv1.ClusterTopologyReconciledHookBlockingV1Beta2Reason,
+			// TODO: Add a protection for messages continuously changing leading to Cluster object changes/reconcile.
+			Message: s.HookResponseTracker.AggregateMessage(),
```
Comment (fabriziopandini) [inline internal/controllers/topology/cluster/conditions.go:156]: Wondering if we should have a sort of stale warning also here

Comment (sbueringer) [reply internal/controllers/topology/cluster/conditions.go:156]: At the moment I only know which hooks are blocking the current reconcile.

I don't know which hook was blocking the previous reconciles. This has two dimensions: the kind of hook BeforeUpgrade, ... but also there could be multiple hooks e.g. someone could register 5 BeforeUpgrade hooks)

File: api/v1beta1/cluster_types.go (hunk 1 above)
Comment (sbueringer) [reply api/v1beta1/cluster_types.go:69]: Reframing my question. Do we want to only change the priority of the TopologyReconciled condition or do we have to do more?

Comment (sbueringer) [reply api/v1beta1/cluster_types.go:69]: Implemented the priority thing. Seems okay to me

Comment (sbueringer) [reply api/v1beta1/cluster_types.go:69]: Hm actually not sure if it's enough, maybe we need an overrideCondition, but PTAL. I added 4 test cases to the Available condition test, that should help to figure out what we want :)

File: internal/controllers/topology/cluster/conditions.go (hunk 2 above)
Comment (fabriziopandini) [reply internal/controllers/topology/cluster/conditions.go:156]: Got it, I was thinking to look at the transition timestamp of the condition, but might be it is not accurate.
Let just keep this in the wish list for now

File: api/v1beta1/cluster_types.go (hunk 1 above)
Comment (fabriziopandini) [reply api/v1beta1/cluster_types.go:69]: I think changing the priority is enough for now.
We can iterate when we go through the refinement phase next weeks / we see this condition together with all the other new conditions

File: internal/controllers/topology/cluster/conditions.go (hunk 2 above)
Comment (sbueringer) [reply internal/controllers/topology/cluster/conditions.go:156]: Ah, I see. I wasn't thinking about that :)

**Comment by fabriziopandini** [comment]
Thanks for adding extensive test coverage!
/lgtm
/approve

**Pull request merged**